ext/dom/xpath.c

```diff
@@ -32,179 +32,85 @@
-		zend_type_error("Handler name must be a string");
-		xmlXPathFreeObject(obj);
-		goto cleanup_no_callable;
-	}
-	ZVAL_STRING(&fci.function_name, (char *) obj->stringval);
-	xmlXPathFreeObject(obj);
-
-	fci.object = NULL;
-	fci.named_params = NULL;
-	fci.retval = &retval;
-
-	if (!zend_make_callable(&fci.function_name, &callable)) {
-		zend_throw_error(NULL, "Unable to call handler %s()", ZSTR_VAL(callable));
-		goto cleanup;
-	} else if (intern->registerPhpFunctions == 2 && zend_hash_exists(intern->registered_phpfunctions, callable) == 0) {
-		zend_throw_error(NULL, "Not allowed to call handler '%s()'.", ZSTR_VAL(callable));
-		goto cleanup;
+static void dom_xpath_ext_function_php(xmlXPathParserContextPtr ctxt, int nargs, php_dom_xpath_nodeset_evaluation_mode evaluation_mode) /* {{{ */
+{
+	dom_xpath_object *intern = dom_xpath_ext_fetch_intern(ctxt);
+	if (!intern) {
+		php_dom_xpath_callbacks_clean_argument_stack(ctxt, nargs);
 	} else {
-		result = zend_call_function(&fci, NULL);
-		if (result == SUCCESS && Z_TYPE(retval) != IS_UNDEF) {
-			if (Z_TYPE(retval) == IS_OBJECT && instanceof_function(Z_OBJCE(retval), dom_node_class_entry)) {
-				xmlNode *nodep;
-				dom_object *obj;
-				if (intern->node_list == NULL) {
-					intern->node_list = zend_new_array(0);
-				}
-				Z_ADDREF(retval);
-				zend_hash_next_index_insert(intern->node_list, &retval);
-				obj = Z_DOMOBJ_P(&retval);
-				nodep = dom_object_get_node(obj);
-				valuePush(ctxt, xmlXPathNewNodeSet(nodep));
-			} else if (Z_TYPE(retval) == IS_FALSE || Z_TYPE(retval) == IS_TRUE) {
-				valuePush(ctxt, xmlXPathNewBoolean(Z_TYPE(retval) == IS_TRUE));
-			} else if (Z_TYPE(retval) == IS_OBJECT) {
-				zend_type_error("A PHP Object cannot be converted to a XPath-string");
-				return;
-			} else {
-				zend_string *str = zval_get_string(&retval);
-				valuePush(ctxt, xmlXPathNewString((xmlChar *) ZSTR_VAL(str)));
-				zend_string_release_ex(str, 0);
-			}
-			zval_ptr_dtor(&retval);
-		}
-	}
-cleanup:
-	zend_string_release_ex(callable, 0);
-	zval_ptr_dtor_nogc(&fci.function_name);
-cleanup_no_callable:
-	if (fci.param_count > 0) {
-		for (i = 0; i < nargs - 1; i++) {
-			zval_ptr_dtor(&fci.params[i]);
-		}
-		efree(fci.params);
+		php_dom_xpath_callbacks_call_php_ns(&intern->xpath_callbacks, ctxt, nargs, evaluation_mode, &intern->dom, dom_xpath_proxy_factory);
 	}
 }
 /* }}} */
 
 static void dom_xpath_ext_function_string_php(xmlXPathParserContextPtr ctxt, int nargs) /* {{{ */
 {
-	dom_xpath_ext_function_php(ctxt, nargs, 1);
+	dom_xpath_ext_function_php(ctxt, nargs, PHP_DOM_XPATH_EVALUATE_NODESET_TO_STRING);
 }
 /* }}} */
 
 static void dom_xpath_ext_function_object_php(xmlXPathParserContextPtr ctxt, int nargs) /* {{{ */
 {
-	dom_xpath_ext_function_php(ctxt, nargs, 2);
+	dom_xpath_ext_function_php(ctxt, nargs, PHP_DOM_XPATH_EVALUATE_NODESET_TO_NODESET);
 }
 /* }}} */
 
+static void dom_xpath_ext_function_trampoline(xmlXPathParserContextPtr ctxt, int nargs)
```

Comment:
I'm confused, why do you need a special version for trampolines?

Reply:
This isn't for zend trampolines, this is a trampoline in the sense that it sits between xpath and php

Reply:
Right, makes sense :)